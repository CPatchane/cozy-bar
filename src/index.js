/* global __TARGET__, __VERSION__ */

'use strict'

import 'babel-polyfill'

import i18n from './lib/i18n'
import stack from './lib/stack'

import BarView from './components/Bar'

const APP_SELECTOR = '[role=application]'

const createElement = function CozyBarCreateElement () {
  const barNode = document.createElement('div')
  barNode.setAttribute('id', 'coz-bar')
  barNode.setAttribute('role', 'banner')
  barNode.classList.add(`coz-target--${__TARGET__}`)

  return barNode
}

const injectDOM = function CozyBarInjectDOM (data) {
  if (document.getElementById('coz-bar') !== null) { return }

  require('./styles/index.css')

  const barNode = createElement()
  const appNode = document.querySelector(APP_SELECTOR)
  if (!appNode) {
    return console.warn(`Cozy-bar is looking for a "${APP_SELECTOR}" tag that contains your application and can't find it :'(â€¦ The BAR is now disabled`)
  }

  document.body.insertBefore(barNode, appNode)

  return new BarView({
    target: barNode,
    data
  })
}

const bindEvents = function CozyBarBindEvents () {
  this._clickOutsideListener = () => this.fire('clickOutside')
  document.body.addEventListener('click', this._clickOutsideListener)

  this._drawerObserver = this.observe('drawerVisible', drawerVisible => {
    document.querySelector('[role=banner]').dataset.drawerVisible = drawerVisible
  })
}

const unbindEvents = function CozyBarUnbindEvents () {
  document.body.removeEventListener('click', this._clickOutsideListener)
  this._drawerObserver.cancel()
}

const getDefaultStackURL = function GetDefaultCozyURL () {
  return document.querySelector('[role=application]').dataset.cozyDomain
}

const getDefaultToken = function GetDefaultToken () {
  return document.querySelector('[role=application]').dataset.cozyToken
}

const getDefaultLang = function GetDefaultLang () {
  return document.documentElement.getAttribute('lang') || 'en'
}

const getDefaultIcon = function GetDefaultIcon () {
  const linkNode = document.querySelector('link[rel="icon"][sizes^="32"]')
  if (linkNode !== null) {
    return linkNode.getAttribute('href')
  } else {
    return 'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7'
  }
}

const init = function CozyBarInit ({
  lang = getDefaultLang(),
  appName,
  iconPath = getDefaultIcon(),
  cozyURL = getDefaultStackURL(),
  token = getDefaultToken()
} = {}) {
  i18n(lang)
  stack.init({cozyURL, token})
  const view = injectDOM({lang, appName, iconPath})

  if (view) {
    bindEvents.call(view)
    view.on('teardown', unbindEvents.bind(view))
  }
}

module.exports = { init, version: __VERSION__ }
