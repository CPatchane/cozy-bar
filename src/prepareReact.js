// Check for global React or load dependency
const ReactBundled = require('react')
const ReactDOMBundled = require('react-dom')

if (!window.React || !window.ReactDOM) {
  console.info(
    'CozyBar: No react related frameworks found, the bar will use its own ones.'
  )
  window.React = ReactBundled
  window.ReactDOM = ReactDOMBundled
}
