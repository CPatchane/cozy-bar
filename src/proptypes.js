import PropTypes from 'prop-types'

export const appShape = PropTypes.shape({
  icon: PropTypes.shape({
    cached: PropTypes.object,
    src: PropTypes.string
  }),
  slug: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  namePrefix: PropTypes.string,
  comingSoon: PropTypes.bool,
  href: PropTypes.string
})
