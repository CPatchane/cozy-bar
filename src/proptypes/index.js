import PropTypes from 'prop-types'

export const appShape = PropTypes.shape({
  slug: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  namePrefix: PropTypes.string,
  comingSoon: PropTypes.bool,
  href: PropTypes.string,
  links: PropTypes.shape({ icon: PropTypes.string.isRequired })
})
