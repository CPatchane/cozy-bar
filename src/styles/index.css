@import "base.css";
@import "indicators.css";
@import "bar.css";
@import "navigation.css";
@import "navigation_section.css";
@import "navigation_item.css";
@import "storage.css";
@import "drawer.css";
