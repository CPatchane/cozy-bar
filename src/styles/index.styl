@import 'base.css'
@import 'indicators.css'
@import 'bar.css'
@import 'nav.css'
@import 'navigation_section.css'
@import 'navigation_item.css'
@import 'storage.css'
@import 'drawer.css'
@import 'claudy.css'
@import 'searchbar.css'
@import 'mobile_app.css'

@import 'ui-base/palette'
@import 'ui-base/mixins'
@import 'ui-base/z-index'

@import 'ui-components/button.styl'

:root {
  --z-index-bar: $bar-index;
  --z-index-selection: $selection-index;
  --z-index-over-selection: $selection-index + 1;
}

[role=banner] {
  --charcoal-grey: charcoal-grey;
  --silver: silver;
  --pale-grey: pale-grey;
  --cool-grey: cool-grey;
  --slate-grey: slate-grey;
  --dodger-blue: dodger-blue;
  --science-blue: science-blue;
  --pomegranate: pomegranate;
}

.coz-bar-hidden
  @extend $hidden

.coz-btn-regular
  @extend $button
  @extend $button--regular
  // built-in stylus em() from cozy-ui is not correcty supported here
  min-height: 40px;
  padding: 13px 15px 11px;

.coz-btn-close
  @extend $button
  @extend $button--close
  // built-in stylus em() from cozy-ui is not correcty supported here
  width: 32px;
  padding: 0;
  margin: 0;
