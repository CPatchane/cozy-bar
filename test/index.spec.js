import { init } from '../src'
import Enzyme, { mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

Enzyme.configure({ adapter: new Adapter() })

describe('The bar library', function () {
  it('should render correctly the cozy-bar', () => {
    // Set up our document body
    document.body.innerHTML =
      '<div role="application">' +
      '  The application will be here' +
      '</div>'
    const options = {
      appName: 'test-app',
      appEditor: 'cozy',
      lang: 'en',
      iconPath: '',
      cozyURL: 'https://mock.cozy',
      token: 'mock',
      displayOnMobile: true
    }
    const Bar = init(options)
    const mounted = mount(Bar)
    expect(mounted.getElement()).toMatchSnapshot()
  })
})
