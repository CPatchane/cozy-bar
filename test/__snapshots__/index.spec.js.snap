// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The bar library should render correctly the cozy-bar 1`] = `
<Provider
  store={
    BarStore {
      "claudyActions": null,
      "helpLink": "",
      "settingsAppURL": "",
      "settingsData": null,
    }
  }
>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(Symbol.observable): [Function],
      }
    }
  >
    <Connect(I18n)
      dictRequire={[Function]}
    >
      <Wrapper
        appEditor="cozy"
        appName="test-app"
        displayOnMobile={true}
        iconPath=""
        isPublic={false}
        onDrawer={[Function]}
        onLogOut={undefined}
        renewToken={null}
        replaceTitleOnMobile={false}
        userActionRequired={undefined}
      />
    </Connect(I18n)>
  </Provider>
</Provider>
`;
